nestkernel/eprop_archiving_node.h

```diff
@@ -117,21 +199,99 @@ class EpropArchivingNodeRecurrent : public EpropArchivingNode< HistEntryEpropRec
+  double compute_piecewise_linear_surrogate_gradient( const double r,
+    const double v_m,
+    const double v_th_adapt,
+    const double V_th,
+    const double beta,
+    const double gamma );
+
+  //! Compute the surrogate gradient with an exponentially decaying function around the spike time (used, e.g., in
+  //! Shrestha and Orchard, 2018).
+  double compute_exponential_surrogate_gradient( const double r,
+    const double v_m,
+    const double v_th_adapt,
+    const double V_th,
+    const double beta,
+    const double gamma );
+
+  //! Compute the surrogate gradient with a function corresponding to the derivative of a fast sigmoid around the spike
+  //! (used, e.g., in Zenke and Ganguli, 2018).
+  double compute_fast_sigmoid_derivative_surrogate_gradient( const double r,
+    const double v_m,
+    const double v_th_adapt,
+    const double V_th,
+    const double beta,
+    const double gamma );
+
+  //! Compute the surrogate gradient with an arctan function around the spike time (used, e.g., in Fang et al., 2021).
+  double compute_arctan_surrogate_gradient( const double r,
+    const double v_m,
+    const double v_th_adapt,
+    const double V_th,
+    const double beta,
+    const double gamma );
+
+  //! Create an entry for the given time step at the end of the eprop history.
+  void emplace_new_eprop_history_entry( const long time_step );
+
+  //! Write the given surrogate gradient value to the history at the given time step.
   void write_surrogate_gradient_to_history( const long time_step, const double surrogate_gradient );
 
   //! Update the learning signal in the eprop history entry of the given time step by writing the value of the incoming
   //! learning signal to the history or adding it to the existing value in case of multiple readout neurons.
-  void write_learning_signal_to_history( const long time_step, const double learning_signal );
+  void write_learning_signal_to_history( const long time_step,
+    const double learning_signal,
+    const bool has_norm_step = true );
 
   //! Create an entry in the firing rate regularization history for the current update.
   void write_firing_rate_reg_to_history( const long t_current_update, const double f_target, const double c_reg );
 
+  //! Calculate the current firing rate regularization and add the value to the learning signal.
+  void write_firing_rate_reg_to_history( const long t,
+    const double z,
+    const double f_target,
+    const double kappa,
+    const double c_reg );
+
   //! Get an iterator pointing to the firing rate regularization history of the given time step.
   std::vector< HistEntryEpropFiringRateReg >::iterator get_firing_rate_reg_history( const long time_step );
 
   //! Return learning signal from history for given time step or zero if time step not in history
-  double get_learning_signal_from_history( const long time_step );
+  double get_learning_signal_from_history( const long time_step, const bool has_norm_step = true );
```

Comment:
See comment on base class.

Reply:
Fixed in https://github.com/nest/nest-simulator/pull/3207/commits/102dc9099b459097152dafa06ccedb18a1ad113b